M2/Macaulay2/packages/HomologicalShiftIdeals.m2

```diff
@@ -0,0 +1,1001 @@
+  Outputs
+    l:List
+      the bounding multidegree of $I$: $\textbf{deg}(I)$
+  Description
+    Text
+      The bounding multidegree of a monomial ideal $I$ of a polynomial ring $S=K[x_1,\ldots,x_n]$ is the vector $\textbf{deg}(I)=(\text{deg}_{x_1}(I),\text{deg}_{x_2}(I),\dots,\text{deg}_{x_n}(I))$ with $$\text{deg}_{x_i}(I)=\max_{u\in G(I)}\max\{j:x_i^j\ \text{divides}\ u\},$$ where $G(I)$ is the minimal monomial generating set of $I$.
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      boundingMultidegree(I)
+///);
+
+boundingMultidegree = method(TypicalValue=>List)
+boundingMultidegree (Ideal) := I -> (
+	if not isMonomialIdeal(I) then error
+        else S:=ring I;
+        n:=numgens S-1;
+        L:=apply(flatten entries mingens I,x->toMultidegree(x));
+        bDeg:={};
+        for i from 0 to n do (
+              bDeg=append(bDeg,max(apply(L,x->x#i)));
+        );
+        bDeg
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the kth multigraded shifts of a monomial ideal
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    multigradedShifts
+    (multigradedShifts,Ideal,ZZ)
+  Headline
+    compute the kth multigraded shifts of a monomial ideal
+  Usage
+    multigradedShifts(I,k)
+  Inputs
+    I:Ideal
+      a monomial ideal $I\subset S=K[x_1,\dots,x_n]$, $K$ a field
+    k:ZZ
+      an integer $k$
+  Outputs
+    l:List
+      the $k$th multigraded shifts of $I$
+  Description
+    Text
+      Given a monomial ideal $I$ of a polynomial ring $S=K[x_1,\ldots,x_n]$ with minimal multigraded free resolution $$\mathbb{F}:\cdots\rightarrow\bigoplus_{j=1}^{\beta_k(I)}S(-\textbf{a}_{k,j})\rightarrow\bigoplus_{j=1}^{\beta_{k-1}(I)}S(-\textbf{a}_{k-1,j})\rightarrow\cdots\rightarrow\bigoplus_{j=1}^{\beta_0(I)}S(-\textbf{a}_{0,j})\rightarrow I\rightarrow 0,$$ the set of the $k\text{th}$ multigraded shifts of $I$ is $$\{\textbf{x}^{\textbf{a}_{k,j}}:j=1,\dots,\beta_k(I)\}.$$
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      multigradedShifts(I,0)
+      multigradedShifts(I,1)
+      multigradedShifts(I,2)
+      multigradedShifts(I,3)
+      multigradedShifts(I,4)
+      multigradedShifts(I,5)
+///);
+
+multigradedShifts = method(TypicalValue=>List)
+multigradedShifts (Ideal,ZZ) := (I,k) -> (
+        if not isMonomialIdeal(I) then error
+        else S:=ring I;
+        n:=numgens S-1;
+        r:=k+1;
+        L:={};
+        for i from 0 to n do (
+              mdeg:={};
+              for j from 0 to n do (
+                     if j==i then mdeg=append(mdeg,1)
+                     else mdeg=append(mdeg,0);
+              );
+              L=append(L,mdeg);
+        );
+        R:=newRing(S,Degrees=>L);
+        f:=map(R,S);
```

Comment:
This is the same map that was constructed for `toMultidegree`, but you're applying it to an ideal instead of a ring element.  Perhaps you could extract this code into a helper function that both methods could call?  (DRY principle when programming -- don't repeat yourself!)